neural_modelling/src/neuron/plasticity/stdp/synapse_dynamics_stdp_mad_neuromodulated_impl.c

```diff
@@ -0,0 +1,455 @@
+//
+// * Maximum value of dendritic delay is 15 (with axonal delay of 0) - It requires 4 bits
+// * Maximum value of axonal delay is 7 (with dendritic delay of 8) - It requires 3 bits
+//
+//             |        Axonal delay       |  Dendritic delay   |       Type        |      Index         |
+//             |---------------------------|--------------------|-------------------|--------------------|
+// Bit count   | SYNAPSE_AXONAL_DELAY_BITS | SYNAPSE_DELAY_BITS | SYNAPSE_TYPE_BITS | SYNAPSE_INDEX_BITS |
+//             |                           |                    |        SYNAPSE_TYPE_INDEX_BITS         |
+//             |---------------------------|--------------------|----------------------------------------|
+#ifndef SYNAPSE_AXONAL_DELAY_BITS
+  #define SYNAPSE_AXONAL_DELAY_BITS 3
+#endif
+
+#define SYNAPSE_AXONAL_DELAY_MASK ((1 << SYNAPSE_AXONAL_DELAY_BITS) - 1)
+
+#define SYNAPSE_DELAY_TYPE_INDEX_BITS \
+    (SYNAPSE_DELAY_BITS + SYNAPSE_TYPE_INDEX_BITS)
+
+#if (SYNAPSE_DELAY_TYPE_INDEX_BITS + SYNAPSE_AXONAL_DELAY_BITS) > 16
+  #error "Not enough bits for axonal synaptic delay bits"
+#endif
+
+//---------------------------------------
+// Structures
+//---------------------------------------
+typedef struct
+{
+  pre_trace_t prev_trace;
+  uint32_t prev_time;
+} pre_event_history_t;
+
+post_event_history_t *post_event_history;
+uint32_t weight_update_constant_component;
+uint32_t last_dopamine_spike_time = 0;
+weight_state_t weight_state;
+
+
+//---------------------------------------
+// Dopamine trace is a simple decaying trace similarly implemented as pre and
+// post trace.
+static inline post_trace_t add_dopamine_spike(
+        uint32_t time, int16_t concentration, int16_t* last_dopamine_level) {
+
+    // Get time since last dopamine spike
+    uint32_t delta_time = time - last_dopamine_spike_time;
+
+    // Apply exponential decay to get the current value
+    int32_t decayed_trace = STDP_FIXED_MUL_16X16(*last_dopamine_level,
+            DECAY_LOOKUP_TAU_D(delta_time));
+
+    // Increase dopamine level due to new spike
+    int16_t new_trace = decayed_trace + concentration;
+
+    last_dopamine_spike_time = time;
+    *last_dopamine_level = new_trace;
+
+    // Return decayed dopamine trace
+    return (post_trace_t) { .stdp_post_trace = 0, .dopamine = new_trace };
+}
+
+static inline void correlation_apply_post_spike(
+        uint32_t time, post_trace_t trace, uint32_t last_pre_time,
+        pre_trace_t last_pre_trace, uint32_t last_non_dopamine_spike_time,
+        uint32_t last_update_time, plastic_synapse_t *previous_state,
+        post_event_history_t *post_event_history) {
+
+    use(&trace);
+
+    // Calculate EXP components in JK's weight update equation
+    int16_t decay_eligibility_trace = DECAY_LOOKUP_TAU_C(
+        time - last_non_dopamine_spike_time);
+    int16_t decay_dopamine_trace = DECAY_LOOKUP_TAU_D(
+        time - post_event_history -> last_dopamine_spike_time);
+    int16_t third_exp_component;
```

Comment:
You also shouldn't use 16-bit integers to STORE stuff that's going to live in registers or on the stack - ARM registers are 32-bit AND it results in bit masking operations whenever you access them